File: scripts/check_command_count_messaging.py
#!/usr/bin/env python3
"""
Command-count messaging gate (regression guard for road-to-pr-34-followups 1.2).

Public surfaces (README.md, AGENTS.md, docs/getting-started.md) advertise
the size of the command catalog. PR #34 collapses atomic commands into
clusters via deprecation shims — the externally meaningful number is the
**active** command count (non-shim files), not the raw file count. This
gate sources canonical counts from `.agent-src.uncondensed/commands/`
frontmatter and fails when any documented number drifts from those.

Canonical counts:
    total   = number of *.md files under .agent-src.uncondensed/commands/
    shims   = files whose frontmatter declares `superseded_by:`
    active  = total - shims

Patterns checked (per file):

    README.md
      hero badge   "/badge/Commands-{N}-…"                    → active
      (Prose phrasings "Browse all {N} active commands" and
       "{N} native commands" were retired in the modernized
       README — the badge alone now carries the count.)

    AGENTS.md
      tree         "commands/  ({N} files — {A} active + {S} deprecation shims)"
      (Thin-Root: only checked when a `commands/` tree block exists.)

    docs/getting-started.md
      browse line  "Browse all {N} active commands"           → active

Exit codes: 0 = clean, 1 = drift detected.
"""

from __future__ import annotations

import re
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from _lib.agent_src import artefact_roots  # noqa: E402

QUIET = "--quiet" in sys.argv

ROOT = Path(__file__).resolve().parent.parent
README = ROOT / "README.md"
AGENTS = ROOT / "AGENTS.md"
GETTING_STARTED = ROOT / "docs" / "getting-started.md"

FM_RE = re.compile(r"^---\s*\n(.*?)\n---", re.DOTALL)
SUPERSEDED_RE = re.compile(r"^superseded_by:\s*\S", re.MULTILINE)


def _command_files() -> list[Path]:
    """Every command ``*.md`` file across all source roots (legacy + packages/*).

    Multi-root aware per ADR-017: post-move the commands live under
    ``packages/<pack>/.agent-src.uncondensed/commands/``, and the
    canonical count is the union across packs (deduped by logical path).
    """
    seen: dict[str, Path] = {}
    for root in artefact_roots():
        cmd_dir = root / "commands"
        if not cmd_dir.is_dir():
            continue
        for f in cmd_dir.rglob("*.md"):
            if f.name == "AGENTS.md":
                continue
            rel = f.relative_to(cmd_dir).as_posix()
            seen.setdefault(rel, f)
    return sorted(seen.values())


def canonical_counts() -> tuple[int, int, int]:
    files = _command_files()
    if not files:
        print("❌  no commands/ directory found under any artefact root", file=sys.stderr)
        sys.exit(1)
    total = shims = 0
    for f in files:
        total += 1
        m = FM_RE.match(f.read_text(encoding="utf-8"))
        fm = m.group(1) if m else ""
        if SUPERSEDED_RE.search(fm):
            shims += 1
    return total, shims, total - shims


def _check(path: Path, pattern: str, expected: int, label: str) -> str | None:
    if not path.exists():
        return f"missing file: {path.relative_to(ROOT)}"
    m = re.search(pattern, path.read_text(encoding="utf-8"))
    if not m:
        return f"{path.relative_to(ROOT)}: pattern not found for `{label}` — /{pattern}/"
    found = int(m.group(1))
    if found != expected:
        return f"{path.relative_to(ROOT)}: `{label}` says {found}, expected {expected}"
    return None


def main() -> int:
    total, shims, active = canonical_counts()
    print(f"Canonical counts: {total} files · {shims} shims · {active} active")

    checks = [
        # README.md — modernized: badge is the sole count surface
        (README, r"/badge/Commands-(\d+)-", active, "hero badge"),
        # docs/getting-started.md — still carries the prose browse line
        (GETTING_STARTED, r"Browse all (\d+) active commands", active, "browse line"),
    ]
    # Shim-specific messaging only applies during a deprecation window.
    # When shims == 0 the clauses are dropped from public docs entirely;
    # re-add these patterns when a new deprecation cycle starts.
    #
    # AGENTS.md is Thin-Root (per agents-md-thin-root skill) — it carries
    # pointers, not an inventory tree. Tree-shaped shim messaging only
    # applies when AGENTS.md actually contains a `commands/` tree block
    # (legacy form). README absorbs the deprecation advertisement either way.
    if shims > 0:
        checks.extend([
            (README, r"\((\d+) files total ", total, "browse meta · total files"),
            (README, r"— (\d+) are deprecation shims", shims, "browse meta · shims"),
        ])
        agents_text = AGENTS.read_text(encoding="utf-8") if AGENTS.exists() else ""
        if re.search(r"commands/\s+\(", agents_text):
            checks.extend([
                (AGENTS, r"commands/\s+\((\d+) files —", total, "tree · total files"),
                (AGENTS, r"files — (\d+) active", active, "tree · active"),
                (AGENTS, r"active \+ (\d+) deprecation shims", shims, "tree · shims"),
            ])

    errors: list[str] = []
    for path, pattern, expected, label in checks:
        err = _check(path, pattern, expected, label)
        if err:
            errors.append(err)

    if not errors:
        if not QUIET:
            print("✅  All command-count messaging in sync with registry.")
        return 0

    print(f"❌  Command-count messaging drift — {len(errors)} mismatch(es):")
    for e in errors:
        print(f"    {e}")
    print(
        "\nFix: update the documented numbers above, or run "
        "`task check-command-count` after editing."
    )
    print(
        "Why this gate exists: see `agents/roadmaps/road-to-pr-34-followups.md` § 1.2."
    )
    return 1


if __name__ == "__main__":
    sys.exit(main())

File: src/scripts/check_command_count_messaging.ts
#!/usr/bin/env tsx
/**
 * Command-count messaging gate (road-to-pr-34-followups 1.2).
 *
 * TypeScript twin of `src/scripts/check_command_count_messaging.py`
 * (ADR-088, Phase 4 / Wave 4c). Mirrors the Python CLI contract EXACTLY —
 * `--quiet` flag (positional anywhere in argv), exit codes (0 clean,
 * 1 drift, 1 no-commands-dir via stderr), stdout/stderr split,
 * byte-identical messages, same canonical-count derivation
 * (iter_commands union), same per-file pattern checks in the same order.
 * No behaviour changes.
 *
 * Sources canonical counts from the command frontmatter and fails when any
 * documented number drifts (README hero badge, getting-started browse line;
 * shim-specific clauses only during a deprecation window).
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';

import { iter_commands } from './_lib/agent_src.js';

const _HERE = fileURLToPath(import.meta.url);
const ROOT = path.resolve(path.dirname(_HERE), '..', '..');
const README = path.join(ROOT, 'README.md');
const AGENTS = path.join(ROOT, 'AGENTS.md');
const GETTING_STARTED = path.join(ROOT, 'docs', 'getting-started.md');

// re.DOTALL, non-greedy: `^---\s*\n(.*?)\n---`.
const FM_RE = /^---\s*\n([\s\S]*?)\n---/;
// re.MULTILINE: `^superseded_by:\s*\S`.
const SUPERSEDED_RE = /^superseded_by:\s*\S/m;

function _exists(p: string): boolean {
    try {
        fs.statSync(p);
        return true;
    } catch {
        return false;
    }
}

function _relTo(target: string): string {
    return path.relative(ROOT, target).split(path.sep).join('/');
}

/** Every command `*.md` across all source layouts, sorted by POSIX path. */
function _command_files(): string[] {
    const out: string[] = [];
    for (const f of iter_commands()) {
        if (path.basename(f) !== 'AGENTS.md') {
            out.push(f);
        }
    }
    return out.sort((a, b) => {
        const ap = a.split(path.sep).join('/');
        const bp = b.split(path.sep).join('/');
        return ap < bp ? -1 : ap > bp ? 1 : 0;
    });
}

function canonical_counts(): [number, number, number] {
    const files = _command_files();
    if (files.length === 0) {
        process.stderr.write('❌  no commands/ directory found under any artefact root\n');
        process.exit(1);
    }
    let total = 0;
    let shims = 0;
    for (const f of files) {
        total += 1;
        const text = fs.readFileSync(f, 'utf-8');
        const m = FM_RE.exec(text);
        const fm = m ? m[1]! : '';
        if (SUPERSEDED_RE.test(fm)) {
            shims += 1;
        }
    }
    return [total, shims, total - shims];
}

interface Check {
    path: string;
    pattern: RegExp;
    // The raw Python regex source string (what the f-string interpolates),
    // kept verbatim so the "pattern not found" message is byte-identical.
    raw: string;
    expected: number;
    label: string;
}

function _check(c: Check): string | null {
    if (!_exists(c.path)) {
        return `missing file: ${_relTo(c.path)}`;
    }
    const text = fs.readFileSync(c.path, 'utf-8');
    const m = c.pattern.exec(text);
    if (!m) {
        return `${_relTo(c.path)}: pattern not found for \`${c.label}\` — /${c.raw}/`;
    }
    const found = parseInt(m[1]!, 10);
    if (found !== c.expected) {
        return `${_relTo(c.path)}: \`${c.label}\` says ${found}, expected ${c.expected}`;
    }
    return null;
}

function main(argv: readonly string[]): number {
    const QUIET = argv.includes('--quiet');
    const [total, shims, active] = canonical_counts();
    process.stdout.write(
        `Canonical counts: ${total} files · ${shims} shims · ${active} active\n`,
    );

    const checks: Check[] = [
        {
            path: README,
            pattern: /\/badge\/Commands-(\d+)-/,
            raw: '/badge/Commands-(\\d+)-',
            expected: active,
            label: 'hero badge',
        },
        {
            path: GETTING_STARTED,
            pattern: /Browse all (\d+) active commands/,
            raw: 'Browse all (\\d+) active commands',
            expected: active,
            label: 'browse line',
        },
    ];
    if (shims > 0) {
        checks.push(
            {
                path: README,
                pattern: /\((\d+) files total /,
                raw: '\\((\\d+) files total ',
                expected: total,
                label: 'browse meta · total files',
            },
            {
                path: README,
                pattern: /— (\d+) are deprecation shims/,
                raw: '— (\\d+) are deprecation shims',
                expected: shims,
                label: 'browse meta · shims',
            },
        );
        const agents_text = _exists(AGENTS) ? fs.readFileSync(AGENTS, 'utf-8') : '';
        if (/commands\/\s+\(/.test(agents_text)) {
            checks.push(
                {
                    path: AGENTS,
                    pattern: /commands\/\s+\((\d+) files —/,
                    raw: 'commands/\\s+\\((\\d+) files —',
                    expected: total,
                    label: 'tree · total files',
                },
                {
                    path: AGENTS,
                    pattern: /files — (\d+) active/,
                    raw: 'files — (\\d+) active',
                    expected: active,
                    label: 'tree · active',
                },
                {
                    path: AGENTS,
                    pattern: /active \+ (\d+) deprecation shims/,
                    raw: 'active \\+ (\\d+) deprecation shims',
                    expected: shims,
                    label: 'tree · shims',
                },
            );
        }
    }

    const errors: string[] = [];
    for (const c of checks) {
        const err = _check(c);
        if (err) {
            errors.push(err);
        }
    }

    if (errors.length === 0) {
        if (!QUIET) {
            process.stdout.write('✅  All command-count messaging in sync with registry.\n');
        }
        return 0;
    }

    process.stdout.write(
        `❌  Command-count messaging drift — ${errors.length} mismatch(es):\n`,
    );
    for (const e of errors) {
        process.stdout.write(`    ${e}\n`);
    }
    process.stdout.write(
        '\nFix: update the documented numbers above, or run ' +
            '`task check-command-count` after editing.\n',
    );
    process.stdout.write(
        'Why this gate exists: see `agents/roadmaps/road-to-pr-34-followups.md` § 1.2.\n',
    );
    return 1;
}

const _isCliEntry =
    process.argv[1] !== undefined &&
    import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href;
if (_isCliEntry || process.argv[1] === _HERE) {
    process.exit(main(process.argv.slice(2)));
}

export {
    ROOT,
    README,
    AGENTS,
    GETTING_STARTED,
    FM_RE,
    SUPERSEDED_RE,
    canonical_counts,
    main,
};
